Src/crypto/chacha20.c

```diff
@@ -0,0 +1,175 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name IMProject nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+// Chacha20 definition document : https://datatracker.ietf.org/doc/html/rfc8439
+
+#include "chacha20.h"
+
+#include "bit_manipulation.h"
+#include "utils.h"
+
+#define KEY_STREAM_SIZE 64U
+#define BLOCK_SIZE      64U
+
+// Hex representation of Magic number: "expand 32-byte k"
+#define CHACHA_MAGIC_NUMBER_PART_1 (0x61707865U)
+#define CHACHA_MAGIC_NUMBER_PART_2 (0x3320646eU)
+#define CHACHA_MAGIC_NUMBER_PART_3 (0x79622d32U)
+#define CHACHA_MAGIC_NUMBER_PART_4 (0x6b206574U)
+
+#define ARRAY_64_ZERO_VALUES { \
+        0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U, \
+        0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U, \
+        0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U, \
+        0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U, \
+        0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U, \
+        0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U, \
+        0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U, \
+        0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U  \
+    }
+
+static inline void quarterround(uint32_t* a, uint32_t* b, uint32_t* c, uint32_t* d);
+static void chacha20_block(const byte_t key[CHACHA20_KEY_SIZE], const uint32_t counter,
+                           const byte_t nonce[CHACHA20_NONCE_SIZE],
+                           byte_t out[KEY_STREAM_SIZE]);
+
+static inline void
+quarterround(uint32_t* a, uint32_t* b, uint32_t* c, uint32_t* d) {
+    *a += *b;
+    *d = BitManipulation_rotl32(*d ^ *a, 16);
+    *c += *d;
+    *b = BitManipulation_rotl32(*b ^ *c, 12);
+    *a += *b;
+    *d = BitManipulation_rotl32(*d ^ *a, 8);
+    *c += *d;
+    *b = BitManipulation_rotl32(*b ^ *c, 7);
+}
+
+static void
+chacha20_block(const byte_t key[CHACHA20_KEY_SIZE], const uint32_t counter, const byte_t nonce[CHACHA20_NONCE_SIZE],

```

Comment:
Name it Block since for other static functions we did not add name of algorithm?
const byte* key?
Remove const from counter, here is the nice article about this:
https://abseil.io/tips/109
const byte_t* nonce?
byte_t* out?